src/InputAreaWidget.ts

```diff
@@ -0,0 +1,270 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+'use strict';
+
+import * as CodeMirror from 'codemirror';
+
+import 'codemirror/lib/codemirror.css';
+
+import 'codemirror/mode/meta';
+
+import * as dmp from 'diff-match-patch';
+
+import {
+  Message
+} from 'phosphor-messaging';
+
+import {
+  IChangedArgs
+} from 'phosphor-properties';
+
+import {
+  ISignal, Signal
+} from 'phosphor-signaling';
+
+import {
+  ResizeMessage, Widget, Panel
+} from 'phosphor-widget';
+
+import {
+  IInputAreaViewModel, ITextEditorViewModel
+} from './InputAreaViewModel';
+
+
+let diffMatchPatch = new dmp.diff_match_patch()
+
+
+/**
+ * A widget which hosts a CodeMirror editor.
+ */
+export
+class CodeMirrorWidget extends Widget {
+
+  /**
+   * Construct a CodeMirror widget.
+   */
+  constructor(model: ITextEditorViewModel) {
+    super();
+    this.addClass('jp-CodeMirrorWidget');
+    this._model = model;
+    this._editor = CodeMirror(this.node);
+    this.updateMimetype(model.mimetype);
+    this.updateLineNumbers(model.lineNumbers);
+    this.updateFixedHeight(model.fixedHeight);
+    this.updateText(model.text);
+
+    this._editor.on('change', (instance, change) => {
+      this._model.text = this._editor.getDoc().getValue();
+    });
+    model.stateChanged.connect(this._modelUpdate, this);
```

Comment:
My convention for naming signal handlers is `onThingSignalName`, so for this case it would be `_onModelStateChanged`. Not saying you should follow this, just letting you know my convention. This let's me look at a method name and immediately make a good guess as to what it's connected to.